File: Two-Pointers/Hard/76-Minimum-Window-Substring/solution.cpp
class Solution {
public:
    string minWindow(string s, string t) {
        //hash table; two pointers;
        //TC:O(n); SC:O(n)
        if(s.size() < t.size()){//corner case
            return "";
        }
        int hash[128];
        fill_n(hash, 128, 0);
        fill_t(t, hash);
        int slow = 0, fast = 0;
        string minString = "";
        int minLen = INT_MAX;
        for(slow = 0; slow < s.size(); slow++){
            while(fast < s.size() && (fast - slow < t.size() || !checkHash(hash))){
                hash[s[fast]]++;
                fast++;
            }
            if(!checkHash(hash)){
                break;
            }
            if(fast - slow < minLen){
                minLen = fast - slow;
                minString = s.substr(slow, fast - slow);
            }
            hash[s[slow]]--;
        }
        return minString;
    }
    
private:
    void fill_t(const string& t, int* hash){
        for(char c : t){
            hash[c]--;
        }
    }
    
    bool checkHash(int* hash){
        for(int i = 0; i < 128; i++){
            if(hash[i] < 0){
                return false;
            }
        }
        return true;
    }
};

File: Two-Pointers/Hard/76-Minimum-Window-Substring/solution.cs
public class Solution {
    public string MinWindow(string s, string t) {
        // two pointers(sliding window); hash table
        //TC:O(n); SC:O(n)
        if(s.Length < t.Length || s.Length == 0 || t.Length == 0) {
            return String.Empty;
        }
        Dictionary<char, int> dictT = new Dictionary<char, int>(); 
        foreach(char c in t) { //store t unique character and appearances in dictT
            dictT[c] = dictT.ContainsKey(c) ? ++dictT[c] : 1;
        }
        int left = 0, right = 0;
        int numOfMatched = 0;
        int[] res = {-1, -1};
        Dictionary<char, int> windowCounts = new Dictionary<char, int>();
        while(right < s.Length) {
            while(right < s.Length) { // find the window
                if(dictT.ContainsKey(s[right])) {
                    windowCounts[s[right]] = windowCounts.ContainsKey(s[right]) ? ++windowCounts[s[right]] : 1;
                    if(dictT[s[right]] == windowCounts[s[right]]) {
                        numOfMatched++;
                    }
                    if(numOfMatched == dictT.Count) {
                        break;
                    }
                }
                right++;
            }
            if(right >= s.Length) {
                break;
            }
            while(numOfMatched == dictT.Count && left <= right - t.Length + 1) { // contract the window
                if(dictT.ContainsKey(s[left])) {
                    windowCounts[s[left]]--;
                    if(windowCounts[s[left]] < dictT[s[left]]) {
                        numOfMatched--;
                    }
                }
                left++;
            }
            if(res[0] == - 1 || res[1] - res[0] > right - left) {
                res[0] = left - 1;
                res[1] = right;
            }
            right++;
        }
        if(res[0] == -1) {
            return String.Empty;
        }
        return s.Substring(res[0], res[1] - res[0] + 1);
    }
}